source/common/network/utility.cc

```diff
@@ -88,24 +88,63 @@ Api::IoCallUint64Result receiveMessage(uint64_t max_rx_datagram_size, Buffer::In
   return result;
 }
 
-} // namespace
-
-Address::InstanceConstSharedPtr Utility::parseInternetAddressNoThrow(const std::string& ip_address,
-                                                                     uint16_t port, bool v6only) {
+StatusOr<sockaddr_in> parseV4Address(const std::string& ip_address, uint16_t port) {
   sockaddr_in sa4;
   memset(&sa4, 0, sizeof(sa4));
-  if (inet_pton(AF_INET, ip_address.c_str(), &sa4.sin_addr) == 1) {
-    sa4.sin_family = AF_INET;
-    sa4.sin_port = htons(port);
-    return instanceOrNull(Address::InstanceFactory::createInstancePtr<Address::Ipv4Instance>(&sa4));
+  if (inet_pton(AF_INET, ip_address.c_str(), &sa4.sin_addr) != 1) {
+    return absl::FailedPreconditionError("failed parsing ipv4");
   }
+  sa4.sin_family = AF_INET;
+  sa4.sin_port = htons(port);
+  return sa4;
+}
+
+StatusOr<sockaddr_in6> parseV6Address(const std::string& ip_address, uint16_t port) {
   sockaddr_in6 sa6;
   memset(&sa6, 0, sizeof(sa6));
-  if (inet_pton(AF_INET6, ip_address.c_str(), &sa6.sin6_addr) == 1) {
+  const auto scope_pos = ip_address.rfind('%');
+  if (scope_pos == std::string::npos) {
+    // Parse IPv6 with no scope.
+    if (inet_pton(AF_INET6, ip_address.c_str(), &sa6.sin6_addr) != 1) {
+      return absl::FailedPreconditionError("failed parsing ipv6 with no scope");
+    }
     sa6.sin6_family = AF_INET6;
-    sa6.sin6_port = htons(port);
+  } else {
+    // Parse IPv6 with scope using getaddrinfo().
+    struct addrinfo hints;
+    memset(&hints, 0, sizeof(hints));
+    struct addrinfo* res = nullptr;
+    // Suppresses any potentially lengthy network host address lookups and inhibit the invocation of
+    // a name resolution service.
+    hints.ai_flags = AI_NUMERICHOST | AI_NUMERICSERV;
+    hints.ai_family = AF_INET6;
+    // Hint that getaddrinfo() need not return a linked list of answers.
+    hints.ai_socktype = SOCK_DGRAM;
+    hints.ai_protocol = IPPROTO_UDP;
+    if (getaddrinfo(ip_address.c_str(), /*service=*/nullptr, &hints, &res) != 0) {
```

Comment:
You probably want https://github.com/envoyproxy/envoy/blob/d93afd72e5b861ea8edcd97ebc26417d61f758c3/envoy/api/os_sys_calls.h#L285

Reply:
I can't remember all the details, but I thought some platforms didn't have great support for `getaddrinfo`. But I might be confusing with `getaddrinfo_a`.

Reply:
Looks like the parsing between linux and windows is different and some basic parsing tests that are passing on linux are actually failing on windows. As suggested I'm following up with @wrowe for some suggestion on how to move forward here.  One option would be to go back to the first implementation where we use the inet_pton for everything and we fallback to the `getaddrinfo` only for IP parsing that has the zone specified. I will follow up once there is new information.